fvm/transactionInvoker.go

```diff
@@ -296,3 +212,113 @@ func (i TransactionInvoker) normalExecution(
 
 	return
 }
+
+// Clear changes and try to deduct fees again.
+func (i TransactionInvoker) errorExecution(
+	proc *TransactionProcedure,
+	txnState *state.TransactionState,
+	nestedTxnId state.NestedTransactionId,
+	env environment.Environment,
+	errs *errors.ErrorsCollector,
+) error {
+	txnState.DisableAllLimitEnforcements()
+	defer txnState.EnableAllLimitEnforcements()
+
+	// log transaction as failed
+	env.Logger().Info().
+		Err(errs.ErrorOrNil()).
+		Msg("transaction executed with error")
+
+	env.Reset()
+
+	// drop delta since transaction failed
+	restartErr := txnState.RestartNestedTransaction(nestedTxnId)
+	if errs.CollectShouldEarlyExit(restartErr) {
+		return errs.ErrorOrNil()
+	}
+
+	// try to deduct fees again, to get the fee deduction events
+	feesError := i.deductTransactionFees(proc, txnState, env)
+
+	// if fee deduction fails just do clean up and exit
+	if feesError != nil {
+		env.Logger().Info().
+			Err(feesError).
+			Msg("transaction fee deduction executed with error")
+
+		if errs.CollectShouldEarlyExit(feesError) {
+			return errs.ErrorOrNil()
+		}
+
+		// drop delta
+		restartErr = txnState.RestartNestedTransaction(nestedTxnId)
+		if errs.CollectShouldEarlyExit(restartErr) {
+			return errs.ErrorOrNil()
+		}
+	}
+
+	return nil
+}
+
+func (i TransactionInvoker) validateCommit(
+	proc *TransactionProcedure,
+	txnState *state.TransactionState,
+	nestedTxnId state.NestedTransactionId,
+	env environment.Environment,
+	processErr error,
+) error {
+	if txnState.NumNestedTransactions() > 1 {
```

Comment:
invert to reduce nesting.

Reply:
I'll leave this as is for now.  I plan to simply this function (and move procedure result population into `commit()`) after this PR.

Since we now early exit on failures, we no longer need to check for `processErr == nil`.  it's always a programming error.